intro/README.md

```diff
@@ -0,0 +1,683 @@
+A D program can be run with unit tests enabled or disabled.
+
+Unit tests are a special function defined like:
+
+```d
+unittest
+{
+    // test code
+}
+```
+
+Individual tests are specified in the unit test using [assert expressions](https://dlang.org/spec/expression.html#AssertExpression).
+There can be any number of unit test functions in a module, including within struct and class declarations.
+They are executed in lexical order.
+Unit tests, when enabled, are run after all static initialization is complete and before the **main()** function is called.
+
+```d
+struct Sum
+{
+    int add(int x, int y) { return x + y; }
+ 
+    unittest
+    {
+        Sum sum;
+        assert(sum.add(3,4) == 7);
+        assert(sum.add(-2,0) == -2);
+    }
+}
+```
+   
+## Contract Programming
+
+Contracts are a breakthrough technique to reduce the programming effort for large projects. Contracts are the concept of preconditions, postconditions and errors.
+
+### Preconditions/Postconditions
+
+The pre contracts specify the preconditions before a statement is executed. The most typical use of this would be in validating the parameters to a function.
+The post contracts validate the result of the statement.
+The most typical use of this would be in validating the return value of a function and of any side effects it has.
+In D, pre contracts begin with in, and post contracts begin with out.
+They come at the end of the function signature and before the opening brace of the function body.
+
+```d
+int fun(ref int a, int b)
+in (a > 0)
+in (b >= 0, "b cannot be negative!")
+out (r; r > 0, "return must be positive")
+out (; a != 0)
+{
+    // function body
+    return 0;
+}
+ 
+int fun(ref int a, int b)
+in
+{
+    assert(a > 0);
+    assert(b >= 0, "b cannot be negative!");
+}
+out (r)
+{
+    assert(r > 0, "return must be positive");
+    assert(a != 0);
+}
+do
+{
+    // function body
+    return 0;
+}
+```
+
+The two functions are almost identical semantically.
+The expressions in the first are lowered to contract blocks that look almost exactly like the second, except that a separate block is created for each expression in the first, thus avoiding shadowing variable names.
+
+### Invariants
+
+Invariants are used to specify characteristics of a class or struct that must always be true (except while executing a member function).
+For example, a class representing a date might have an invariant that the day must be 1..31 and the hour must be 0..23:
+
+```d
+class Date
+{
+    int day;
+    int hour;
+ 
+    this(int d, int h)
+    {
+        day = d;
+        hour = h;
+    }
+ 
+    invariant
+    {
+        assert(1 <= day && day <= 31);
+        assert(0 <= hour && hour < 24, "hour out of bounds");
+    }
+}
+```
+
+For public or exported functions, the order of execution is:
+1. preconditions
+2. invariant
+3. function body
+4. invariant
+5. postconditions
+
+### Quiz
+
+What is the output of the following code?
```

Comment:
In addition, "what is the output of this program" is not a very relevant quiz question in general. We don't want to teach students to be human compilers.